teres/bkr_handlers.py

```diff
@@ -201,6 +206,79 @@ class ThinBkrHandlerError(teres.HandlerError):
     pass
 
 
+class _IncrementalUploader:
+    """
+    Uploader utility object that will keep track of previous uploads and
+    try to detect when a file has grown since last upload and upload just
+    the new bytes.
+
+    Note that this detection is very naiive.  For files that might change
+    in any other way than appending new bytes, you probably only want to
+    make sure to only use upload_whole().
+
+    The detection also does not respect file path; file identity is tracked
+    only in terms of URL.  (That is, uploading file to different URL's will
+    upload twice, but uploading different files to same URL will probably
+    break the detection.)
+    """
+
+    def __init__(self):
+        self._next_chunk_pos = {}
+
+    def upload_chunk(self, handle, url):
+        """
+        Upload increment (added bytes) from file-like *handle* a to URL *url*.
+
+        Based on previous uploads (either invoked by this method or upload_whole()),
+        read only new bytes and "remember" the upload.
+        """
+        if url not in self._next_chunk_pos:
+            # new file (0 size is ok)
+            self.upload_whole(handle, url)
+            return
+        range_from = self._next_chunk_pos[url]
+        payload, range_from, range_to = self._safe_read(handle, range_from)
+        if not payload:
+            logger.info("ThinBkrHandler: nothing new to upload: 0 bytes for %s" % url)
+            return
+        logger.info("ThinBkrHandler: uploading file chunk: %d bytes to %s" % (len(payload), url))
+        headers = {'Content-Range': 'bytes %d-%d/*' % (range_from, range_to)}
+        http_put(url, payload, **headers)
+        self._next_chunk_pos[url] = range_to + 1
+
+    def upload_whole(self, handle, url):
+        """
+        Upload all bytes from file-like *handle* to URL *url*.
+
+        Keep record of uploaded bytes so that if file grows, subsequent calls
+        to upload_chunk() will upload only the newly added bytes.
+        """
+        payload, range_from, range_to = self._safe_read(handle)
+        if url in self._next_chunk_pos:
+            logger.info("_IncrementalUploader: re-uploading file: %d bytes to %s"
+                        % (len(payload), url))
+        else:
+            logger.info("_IncrementalUploader: uploading new file: %d bytes to %s"
+                        % (len(payload), url))
+        http_put(url, payload)
+        self._next_chunk_pos[url] = range_to + 1
+
+    def _safe_read(self, handle, range_from=0):
```

Comment:
The "safe" may be misleading especially when the method is not thread safe. It also does "destructive" read when reading pipe like file object. I'd avoid using the "safe" word to avoid possible confusion.

Maybe some name like `_tell_read_seek` would be better as that's exactly what it does (like C macro would do). That's just idea.

Reply:
Good point.  I used `_tell_read_seek`.

I've also added a note on the safety/destructivity.

As an aside, the thread safety / destructivity remains an issue in itself but fixing that is outside of scope of this PR. (Some band-aids could be applied but It's rather an architecture issue.)

Reply:
Yep, that's totally ok and understandable. Thanks